Add SoftplusTransform

jacobhepkema opened the pull request:
This pull request introduces `SoftplusTransform` to `torch.distributions.transforms`. `SoftplusTransform` transforms via the mapping `Softplus(x) = log(1 + exp(x))`. Note that the transform is different to [`torch.nn.Softplus`](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html#torch.nn.Softplus), as that has additional `beta` and `threshold` parameters. Inverse and `log_abs_det_jacobian` for a more complex `SoftplusTransform` can be added in the future.

@vitkl @fritzo 

Addresses the issue discussed here: [pyro issue 855](https://github.com/pyro-ppl/numpyro/issues/855)

jacobhepkema committed:
Add SoftplusTransform

SoftplusTransform transforms via the mapping $\text{Softplus}(x) = \log(1 + \exp(x))$. Note that this is different to `torch.nn.functional.softplus`, as that has additional $\beta$ and $threshold$ parameters

jacobhepkema committed:
Update transforms.py

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+    Transform via the mapping :math:`\text{Softplus}(x) = \log(1 + \exp(x))`.
+    """
+    domain = constraints.real_vector
+    codomain = constraints.positive
```
```py
codomain = constraints.independent(constraints.positive, 1)
```
(this would have been caught by unit tests)

jacobhepkema replied:
What does `constraints.independent` do exactly? The test that I added (locally - I will upload soon) to `test/distributions/test_transforms.py` (adding a line `SoftplusTransform(cache_size=cache_size)` in `get_transforms`) will only pass if I specify this as `constraints.positive` (otherwise it gives an `AssertionError` at `assert torch.Size([4, 5]) == torch.Size([4])` in `test_jacobian(transform)`)

fritzo replied:
Oh I'm sorry, I think this should actually assume `.event_dim = 0` via:
```py
domain = constraints.real  # not real_vector
codomain = constraints.positive
```
The `independent(-, 1)` and `real -> real_vector` both say to set `.event_dim = 1` which determines the shape of the `.log_abs_det_jacobian`. If `.event_dim == 0` then `.log_abs_det_jacobian()` is computed element-wise (which is what you want here). If `.event_dim == 1` then the jacobian is computed for the entire `shape[-1:]` in total, but you won't want that here since `SoftplusTransform` can apply to even scalar tensors (`.shape == ()`).

jacobhepkema replied:
Ah thank you - my bad for using `real_vector` :) Updated this now and added the test, thanks a lot for reviewing

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+        return y.expm1().log()
+
+    def log_abs_det_jacobian(self, x, y):
+        return -(1 + (-x).exp()).log()
```
```py
-(-x).exp().log1p()
```

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+        return isinstance(other, SoftplusTransform)
+
+    def _call(self, x):
+        return (1 + x.exp()).log()
```
```py
x.exp().log1p()
```

fritzo reviewed (changes requested):
Thanks for adding this!

Could you please add a test example to the `get_transforms()` function in test/distributions/test_transforms.py, and ensure tests pass? This should be a mere one-line change.

jacobhepkema committed:
Update codomain, call, log_abs_det_jacobian

Updated SoftplusTransform by implementing feedback from https://github.com/pytorch/pytorch/pull/52300#pullrequestreview-591470077

jacobhepkema committed:
Update transforms.py

jacobhepkema committed:
Update test_transforms.py

jacobhepkema committed:
Merge pull request #1 from jacobhepkema/jacobhepkema-patch-1

Added test for `SoftplusTransform(Transform)` to `test/distributions/test_transforms.py` (it passes locally). Updated `SoftplusTransform(Transform)` in `torch/distributions/transforms.py` to reflect comments from here https://github.com/pytorch/pytorch/pull/52300#discussion_r577012561

fritzo reviewed (approved):
LGTM

fehiepsi commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+        return isinstance(other, SoftplusTransform)
+
+    def _call(self, x):
+        return x.exp().log1p()
```
This is not stable for moderate `x` values. It is better to return `softplus(x)`

vitkl replied:
@fehiepsi doesn't using pytorch `softplus` (with `beta=1, threshold=20`) change the derivative of the transform? E.g. when `threshold > 20` `log_abs_det_jacobian` should return 1.

fehiepsi replied:
Do you mean log_abs_det return 0?

Using threshold looks fine to me though it seems to me that it is not needed. Probably, PyTorch doesn't use `logaddexp(0, x)` under the hood, so it needs the threshold...

vitkl replied:
Yes, I mean that log_abs_det should return 0 when x > 20 because when x > 20 the transform is y=x not y=softplus(x). Maybe I am missing something.

fehiepsi replied:
`-softplus(-20.)` returns `tensor(-2.0612e-09)`, which is good enough I guess.

vitkl replied:
Makes sense to use approximation then.

fehiepsi commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+        return y.expm1().log()
+
+    def log_abs_det_jacobian(self, x, y):
+        return -(-x).exp().log1p()
```
A more stable formula would be: `-softplus(-x)`

vitkl replied:
Same here - https://github.com/pytorch/pytorch/pull/52300#discussion_r579817195

fehiepsi commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+        return x.exp().log1p()
+
+    def _inverse(self, y):
+        return y.expm1().log()
```
Probably a more stable version will be:
```
(-y).expm1().neg().log() - y
```
The point is if possible, we might avoid using `exp(x)` for moderate `x`. Here we know that `y` is positive, so `-y` won't be a large number

jacobhepkema replied:
Since [`torch.nn.functional.softplus`](https://pytorch.org/docs/master/nn.functional.html#torch.nn.functional.softplus) reverts to a linear function when `x` is above `20` (default threshold), this won't pass the tests - alternatively, I could use the `softplus` for `log_abs_det_jacobian` and `_call` as you suggested and use this here:
```
torch.where(y > 20, y, y.expm1().log())
```
(will `torch.where` make it slower?)

fehiepsi replied:
Sorry, my mistake, the correct formula should be `(-y).expm1().neg().log() + y`

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -600,6 +601,26 @@ def log_abs_det_jacobian(self, x, y):
+    Transform via the mapping :math:`\text{Softplus}(x) = \log(1 + \exp(x))`.
+    """
+    domain = constraints.real
+    codomain = constraints.positive
```
Could you also add:
```py
bijective = True
sign = +1
```

jacobhepkema committed:
Update transforms.py

jacobhepkema committed:
SoftplusTransform: softplus(x)

`_inverse(y)` is now defined as `torch.where(y > 20, y, y.expm1().log())` as `softplus(x)` reverts to the linear function when `y > 20`.

jacobhepkema committed:
Update transforms.py

jacobhepkema committed:
Update _inverse of SoftplusTransform

Rather than using `torch.where()` now defined as `(-y).expm1().neg().log() + y`

fehiepsi reviewed (approved):
LGTM, thanks @jacobhepkema!

fritzo commented:
@pytorchbot merge this please

(test failures are unrelated: merge conflict in an untouched script file)

fehiepsi commented:
@pytorchbot merge this please

fritzo commented:
@neerajprad any idea what happened to this old PR?

neerajprad commented:
@jacobhepkema: Could you merge your changes / rebase on top of the latest master and push?

neerajprad commented:
@jacobhepkema, @fritzo: This diff is stale unfortunately, and I cannot push to this fork to update it. Happy to close this PR and push to my own fork to get this merged. Let me know.

jacobhepkema commented:
@neerajprad Hi, sorry for the late reply - will push to the latest master today 👍

jacobhepkema committed:
merged and resolved conflict in test/distributions/test_transforms.py to add SoftplusTransform

jacobhepkema committed:
add SoftplusTransform to docs conf.py

albanD commented:
@fritzo same doc question on this one as the other PR. Is it ok not to document?

albanD commented:
@pytorchbot rebase this please
